### PR TITLE
[CALCITE-3448] AggregateOnCalcToAggregateUnifyRule ignores Project…

### PR DESCRIPTION
**Pull request by jinxing64**
… incorrectly when there's missing grouping or mapping breaks ordering (Jin Xing)

CALCITE-3087 fixed AggregateOnProjectToAggregateUnifyRule when mapping under Aggregate breaks ordering. But it fails when there's missing grouping in query compared with target.
Below matching fails:
```
        MV:
        select empid, deptno, name, count(*) from emps
        group by empid, deptno, name

        Query:
        select name, empid, count(*) from emps group by name, empid
```
Note that -- comparing groupings in MV `(empid, deptno, name)` and groupings in query `(name, empid)`, `deptno` is missed and ordering is broken

### CODE REVIEW

**Comment by jinxing64** [comment]
cc @DonnyZone 
I touched your code. Please give a check when you have time :)
Thanks a lot ~

**Comment by DonnyZone** [comment]
Hi @jinxing64, sorry for the late reply! I'm outing these days, and I will take a look tomorrow.

**Comment by jinxing64** [comment]
Thanks a lot @DonnyZone ~

File: core/src/main/java/org/apache/calcite/util/mapping/Mappings.java [hunk 1]
```diff
@@ -434,7 +434,9 @@ public static boolean keepsOrdering(TargetMapping mapping) {
       if (target != -1 && target < prevTarget) {
         return false;
       }
-      prevTarget = target;
+      if (target != -1) {
```
Comment (DonnyZone) [inline core/src/main/java/org/apache/calcite/util/mapping/Mappings.java]: Maybe we can combine two `if` clause together?

**Comment by DonnyZone** [comment]
LGTM!

**Comment by DonnyZone** [comment]
By the way, have you test the below case?
```
Query:
select name, name, count(*) from emps group by name, empid
```

File: core/src/main/java/org/apache/calcite/util/mapping/Mappings.java (hunk 1 above)
Comment (jinxing64) [reply core/src/main/java/org/apache/calcite/util/mapping/Mappings.java]: Sure

**Comment by jinxing64** [comment]
I tested the test case you provided
```
MV:
select empid, deptno, name, count(*) from emps
group by empid, deptno, name
Query:
select name, name, count(*) from emps group by name, empid
```
The result is correct as below:
``` 
EnumerableCalc(expr#0..2=[{inputs}], name=[$t1], name0=[$t1], EXPR$1=[$t2])
   EnumerableAggregate(group=[{0, 2}], EXPR$2=[$SUM0($3)])
     EnumerableTableScan(table=[[hr, m0]])
```

**Comment by jinxing64** [comment]
Hi Zhu Feng ~
Thanks a lot for your review and approval and I updated the PR by your comments

**Comment by jinxing64** [comment]
Hi ~ committers, 
The bug this PR tries to fix seems obvious ~
Could you please help review this ?
It's great if I can get your comments and merge it if possible ~

**Comment by DonnyZone** [comment]
Ping @hsyuan

**Comment by hsyuan** [comment]
Will do

File: core/src/test/java/org/apache/calcite/test/MaterializationTest.java [hunk 2]
```diff
@@ -864,6 +864,22 @@ private void checkNoMaterialize(String materialize, String query,
                 + "    EnumerableTableScan(table=[[hr, m0]])"));
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-3448">[CALCITE-3448]
+   * AggregateOnProjectToAggregateUnifyRule ignores Project incorrectly when
```
Comment (hsyuan) [inline core/src/test/java/org/apache/calcite/test/MaterializationTest.java]: Update rule name

**Comment by jinxing64** [comment]
Thanks a lot for approval ~

**Pull request closed without merging**